### PR TITLE
Working LLVM 6.0 support, without CUDA

### PR DESCRIPTION
**Pull request by ProfFan**
This is a preliminary work on supporting newer LLVM JIT. Ultimately I think MCJIT is going to be replaced by Orc, however this PR enables basic functions without too much work.

Tested on my Mac running macOS High Sierra and LLVM 6.0.1.

### CODE REVIEW

**Comment by ProfFan** [comment]
CUDA support is broken because of already reported issue #288:
```
cuModuleGetFunction: cuda reported error 500: named symbol not found
```

**Comment by ProfFan** [comment]
Here is the test report:
```
=================
= FAILING tests
=================
canon2.t
class2.t
cudaagg.t
cudaaggregate.t
cudaatomic.t
cudaconst2.t
cudaglobal.t
cudahello.t
cudaoffline.t
cudaoo.t
cudaprintf.t
cudashared.t
cudatest.t
cudatex.t
decltwicef.t
defer.t
defercond.t
dynlib.t
fails/abouttofreeze.t
fails/arith.t
fails/attrload.t
fails/attrload2.t
fails/attrstore.t
fails/attrstore2.t
fails/cast.t
fails/comparestruct.t
fails/constructor.t
fails/defer.t
fails/defer1.t
fails/defergoto.t
fails/defergoto2.t
fails/dupvar.t
fails/entrieserror.t
fails/escapeerr.t
fails/escapemacro.t
fails/f0.t
fails/f1.t
fails/f10.t
fails/f11.t
fails/f12.t
fails/f13.t
fails/f14.t
fails/f15.t
fails/f16.t
fails/f17.t
fails/f18.t
fails/f19.t
fails/f2.t
fails/f20.t
fails/f21.t
fails/f22.t
fails/f23.t
fails/f24.t
fails/f25.t
fails/f26.t
fails/f27.t
fails/f28.t
fails/f29.t
fails/f3.t
fails/f30.t
fails/f31.t
fails/f32.t
fails/f33.t
fails/f34.t
fails/f35.t
fails/f36.t
fails/f37.t
fails/f38.t
fails/f39.t
fails/f4.t
fails/f40.t
fails/f41.t
fails/f43.t
fails/f44.t
fails/f45.t
fails/f46.t
fails/f47.t
fails/f48.t
fails/f5.t
fails/f7.t
fails/f8.t
fails/f9.t
fails/failtrunc.t
fails/fnindex.t
fails/fnpointer.t
fails/for.t
fails/for2.t
fails/forlist.t
fails/forlist2.t
fails/forlist3.t
fails/getmethod.t
fails/gettype.t
fails/ifelse.t
fails/ifelse2.t
fails/incomplete.t
fails/intrinsic1.t
fails/intrinsic2.t
fails/intrinsic3.t
fails/intrinsic4.t
fails/intrinsic5.t
fails/let.t
fails/lvaluetreelist.t
fails/macroselect.t
fails/metatype.t
fails/methodantiquote.t
fails/missingmetamethod.t
fails/missingmethod.t
fails/nesterror.t
fails/nilreturn.t
fails/nilsym.t
fails/niltype2.t
fails/opaque.t
fails/opaque2.t
fails/overload.t
fails/overload3.t
fails/overload4.t
fails/quoteescape.t
fails/quoteescape2.t
fails/recfn.t
fails/recpattern.t
fails/recursivecompile.t
fails/recursivecompile2.t
fails/symboltest.t
fails/symbolvar.t
fails/symbolvar2.t
fails/symbolvar3.t
fails/symbolvar6.t
fails/symbolvar7.t
fails/symparam.t
fails/symparam3.t
fails/ucall.t
fails/usercastwrong.t
fails/veclit.t
fails/veclit2.t
fails/veclit3.t
fails/vectorerror.t
fails/vectype.t
foundintermediate2.t
foundintermediate3.t
globals.t
hasbeenfrozen.t
leaktest.t
let2.t
linklibrary.t
luabridge2.t
simpleglobal.t
teststd.t
vars.t
vars2.t
vecarith.t
zeroreturn.t
zeroreturn2.t
=================

369 tests passed. 152 tests failed.
```

**Comment by ProfFan** [comment]
For all the fail cases:
```
fails/veclit.t:1: module 'fail' not found:
	no field package.preload['fail']
	no file 'build/share/luajit-2.0.5/fail.lua'
	no file 'build/share/luajit-2.0.5/fail.so'
	no file './fail.t'
	no file '/Users/proffan/Projects/Development/Robotics/Optlang/terra/release/share/terra/fail.t'
	no internal file '/fail.t'
stack traceback:
	[C]: in function 'require'
	fails/veclit.t:1: in main chunk
```
I believe it's merely a path issue.

**Comment by ProfFan** [comment]
On Travis-CI, the result is:
```
=================
= FAILING tests
=================
defercond.t
zeroreturn.t
luabridge2.t
simpleglobal.t
hasbeenfrozen.t
let2.t
class2.t
vecarith.t
globals.t
defer.t
vars.t
vars2.t
teststd.t
zeroreturn2.t
canon2.t
=================
506 tests passed. 15 tests failed.
```

No idea why currently.

**Comment by ProfFan** [comment]
It appears to be some kind of error when determine the existence of an array global. 
```
globals.t:
local c = terralib.global(int[3])
assert(3 == c:get()[0]) error
testvector.t:
assert(g:get() == 49*50/2) error
```
Time to sleep.

**Comment by ProfFan** [comment]
I did a little more research on the globals issue:
```lua

struct A { a : int }
local a = terralib.global(A)
a:set({3})
local b = terralib.global(true)
local c = terralib.global(int[3])
local d = terralib.global(3.55)

local IO = terralib.includec("stdio.h")

c[0] = 0xdeadbeef
c[1] = 0
c[2] = 0xdeadbeef

terra foobar()
	c[0] = 1337
	c[1] = 1337
	c[2] = 1337
	return a.a + d
end

e = global(int)
terra doihaveaname()
	e = 10
	return e
end

terra foobar_test()
	IO.printf("foobar_test:%#016x\n",c)
	IO.printf("%d\n",c[0])
	IO.printf("%d\n",c[1])
	IO.printf("%d\n",c[2])
	
	return a
end

assert(doihaveaname() == 10)
assert(6.55 == foobar())

assert(true == b:get())
print("HERE")
print(c:get()[0])
print(c:get()[1])
print(c:get()[2])
print(c[0])
print(c[1])
print(c[2])
print(c:get())
print(c:getpointer())
foobar_test()
assert(3 == c:get()[0])

```
Outputs:
```
HERE
2
0
-514838880
3735928559
0
3735928559
cdata<int (&)[3]>: 0x7f8ee1501110
cdata<int (*)[3]>: 0x7f8ee1501110
foobar_test:0x000000102e7040
1337
1337
1337
tests/globals.t:51: assertion failed!
stack traceback:
	[C]: in function 'assert'
	tests/globals.t:51: in main chunk
```
It appears that `c` in terra and LuaJIT points to different things? @elliottslaughter Can you run this snippet on LLVM<=3.8 and post the results using code in this PR?

**Comment by ProfFan** [comment]
@elliottslaughter I believe 02dd8a9 is close to mergeable state :) I tested on an GCE machine with Ubuntu 18.04 and LLVM 6.0, the result:

```
=================
= FAILING tests
=================
classifyfloatstructs.t
=================

520 tests passed. 1 tests failed.
```

**Comment by ProfFan** [comment]
On my Mac with macOS High Sierra and LLVM 6.0.1:
```
=================
= FAILING tests
=================
decltwicef.t
fails/abouttofreeze.t
fails/arith.t
fails/attrload.t
fails/attrload2.t
fails/attrstore.t
fails/attrstore2.t
fails/cast.t
fails/comparestruct.t
fails/constructor.t
fails/defer.t
fails/defer1.t
fails/defergoto.t
fails/defergoto2.t
fails/dupvar.t
fails/entrieserror.t
fails/escapeerr.t
fails/escapemacro.t
fails/f0.t
fails/f1.t
fails/f10.t
fails/f11.t
fails/f12.t
fails/f13.t
fails/f14.t
fails/f15.t
fails/f16.t
fails/f17.t
fails/f18.t
fails/f19.t
fails/f2.t
fails/f20.t
fails/f21.t
fails/f22.t
fails/f23.t
fails/f24.t
fails/f25.t
fails/f26.t
fails/f27.t
fails/f28.t
fails/f29.t
fails/f3.t
fails/f30.t
fails/f31.t
fails/f32.t
fails/f33.t
fails/f34.t
fails/f35.t
fails/f36.t
fails/f37.t
fails/f38.t
fails/f39.t
fails/f4.t
fails/f40.t
fails/f41.t
fails/f43.t
fails/f44.t
fails/f45.t
fails/f46.t
fails/f47.t
fails/f48.t
fails/f5.t
fails/f7.t
fails/f8.t
fails/f9.t
fails/failtrunc.t
fails/fnindex.t
fails/fnpointer.t
fails/for.t
fails/for2.t
fails/forlist.t
fails/forlist2.t
fails/forlist3.t
fails/getmethod.t
fails/gettype.t
fails/ifelse.t
fails/ifelse2.t
fails/incomplete.t
fails/intrinsic1.t
fails/intrinsic2.t
fails/intrinsic3.t
fails/intrinsic4.t
fails/intrinsic5.t
fails/let.t
fails/lvaluetreelist.t
fails/macroselect.t
fails/metatype.t
fails/methodantiquote.t
fails/missingmetamethod.t
fails/missingmethod.t
fails/nesterror.t
fails/nilreturn.t
fails/nilsym.t
fails/niltype2.t
fails/opaque.t
fails/opaque2.t
fails/overload.t
fails/overload3.t
fails/overload4.t
fails/quoteescape.t
fails/quoteescape2.t
fails/recfn.t
fails/recpattern.t
fails/recursivecompile.t
fails/recursivecompile2.t
fails/symboltest.t
fails/symbolvar.t
fails/symbolvar2.t
fails/symbolvar3.t
fails/symbolvar6.t
fails/symbolvar7.t
fails/symparam.t
fails/symparam3.t
fails/ucall.t
fails/usercastwrong.t
fails/veclit.t
fails/veclit2.t
fails/veclit3.t
fails/vectorerror.t
fails/vectype.t
foundintermediate2.t
foundintermediate3.t
leaktest.t
symbolmangling.t
=================
397 tests passed. 124 tests failed.
```
All the `fail/*`s are because of a path issue.

**Comment by ProfFan** [comment]
Fixing the path issue is actually quite easy: `export LUA_PATH="$(pwd)/tests"` does the magic and on my Mac running macOS High Sierra:
```
=================
= FAILING tests
=================
symbolmangling.t
=================

520 tests passed. 1 tests failed.
```

File: .travis.yml [hunk 1]
```diff
@@ -15,7 +15,7 @@ env:
 matrix:
   include:
     - os: linux
-      dist: bionic
+      dist: trusty
```
Comment (elliottslaughter) [inline .travis.yml]: I think we could test this on macOS too by moving the env line up top and then adding some extra downloads to the uname == Darwin section below (just download from llvm.org like we do for 3.8 and 3.5; ok to do it unconditionally).

File: src/llvmheaders.h [hunk 2]
```diff
@@ -57,11 +64,13 @@
 #include "llvmheaders_38.h"
 #elif LLVM_VERSION == 39
 #include "llvmheaders_39.h"
+#elif LLVM_VERSION == 60
+#include "llvmheaders_60.h"
 #else
-#error "unsupported LLVM version"
+//#error "unsupported LLVM version"
```
Comment (elliottslaughter) [inline src/llvmheaders.h]: Can this be uncommented? Are there any problems that occur if it is?

File: src/tcompiler.cpp [hunk 3]
```diff
@@ -400,6 +424,7 @@ int terra_compilerinit(struct terra_State * T) {
         return LUA_ERRRUN;
 #endif
     }
+
```
Comment (elliottslaughter) [inline src/tcompiler.cpp]: Let's clean up the extra newline so we don't have unnecessary diffs floating around.

File: src/tcompiler.cpp [hunk 4]
```diff
@@ -2993,16 +3045,38 @@ static int terra_linkllvmimpl(lua_State * L) {
...
+    #elif LLVM_VERSION >= 60
+    Expected<std::unique_ptr<Module>> mm = parseBitcodeFile(mb.get()->getMemBufferRef(),*TT->ctx);
     #elif LLVM_VERSION >= 37
     ErrorOr<std::unique_ptr<Module>> mm = parseBitcodeFile(mb.get()->getMemBufferRef(),*TT->ctx);
     #else
     ErrorOr<Module *> mm = parseBitcodeFile(mb.get().get(),*TT->ctx);
     #endif
     if(!mm) {
         if(fromstring)
+            #if LLVM_VERSION >= 50
```
Comment (elliottslaughter) [inline src/tcompiler.cpp]: Could we swap the `{` and the `#if`? Having the `if` statement be single-line in only one branch is making me nervous.

File: src/tcompiler.cpp (hunk 3 above)
Comment (ProfFan) [reply src/tcompiler.cpp]: Maybe we can just start using `clang-format`?

File: src/tcompiler.cpp (hunk 4 above)
Comment (ProfFan) [reply src/tcompiler.cpp]: Ahh. That is to have the minimal changes possible. Fixed :)

File: src/llvmheaders.h (hunk 2 above)
Comment (ProfFan) [reply src/llvmheaders.h]: Done.

File: .travis.yml (hunk 1 above)
Comment (ProfFan) [reply .travis.yml]: Fixed. However I cannot find a prebuilt package in the 6.0.1 tree for macOS, so I just used 6.0.0

**Comment by ProfFan** [comment]
About the failing test on macOS:
```
tests/classifyfloatstructs.t
```
I think the debugging mechanism changed somewhat in `OrcMCJITReplacement`, as `C->functioninfo` is always empty, so `terralib.lookupsymbol(add4_val_val,&si)` always return `false`, though I need to dive a little bit more into it to confirm this.

**Comment by ProfFan** [comment]
Nice ref here: https://groups.google.com/forum/#!topic/llvm-dev/B7quHkDRoYk

Suggests that we cannot just call `RegisterJITEventListener` and hope it will work-not possible with non-MCJIT. I think I should try refer to Julia's impl.

**Comment by ProfFan** [comment]
Too strange: if I change the test to something like this, then it works magically:
```
********* passing to lua ************
C = terralib.includecstring [[
    #include <stdio.h>
    #include <stdlib.h>
	#include <stdlib.h>
	#include <math.h>
    #include <sys/time.h>
    double CurrentTimeInSeconds() {
		return time(NULL);
    }
    FILE* testfopen(char* a, char* b) {
        return fopen(a,b);
    }
]]
C.printf("test %s %d\n", "ttt")
C.srand(354)
print(C.CurrentTimeInSeconds())
C.puts("test\n")
f = C.testfopen("./broken.t","r")
-- f = C.fopen("./broken.t","r")
assert(f == nil)
C.printf("test %s\n", "ttt")
*************************************
test ttt 0
Struct Layout Is:
%"type 0x7f896002d840" = type {}
End Layout
1534929125
test

test ttt
deleting function: puts
... uses not empty, removing body but keeping declaration.
... finish delete.
deleting function: srand
... uses not empty, removing body but keeping declaration.
... finish delete.
deleting function: printf
... uses not empty, removing body but keeping declaration.
... finish delete.
```
I personally suspect a parser issue.

**Comment by ProfFan** [comment]
This one also works
```lua
C = terralib.includecstring [[
    typedef struct {
        char *fpos;
        void *base;
        unsigned short handle;
        short flags;
        short unget;
        unsigned long alloc;
        unsigned short buffincrement;
      } FILE;
      
    FILE *fopen(const char *filename, const char *mode);
]]

f = C.fopen("./broken.t","r")
-- f = C.fopen("./broken.t","r")
assert(f == nil)
```

**Comment by ProfFan** [comment]
There has to be something in the mangling process: should be `fopen`, not `_fopen`
```
********* passing to lua ************
C = terralib.includecstring [[
    #include <stdio.h>

    FILE* fopenhhh(const char * __restrict __filename, const char * __restrict __mode){
        return fopen("a","b");
    }
]]
C.fopenhhh("./broken.t","r")
print(C.fopenhhh)
print("Real fopen")
print(C.fopen)
f = C.fopen("./broken.t","r")
-- f = C.fopen("./broken.t","r")
assert(f == nil)
-- C.printf("test %s\n", "ttt")
*************************************
[C]:-1:                 terra fopenhhh :: {&int8,&int8} -> &__sFILE = <extern>

Real fopen
[C]:-1:                 terra _fopen :: {&int8,&int8} -> &__sFILE = <extern>

0   ???                                 0x0000000000000000 0x0 + 0
```

**Comment by ProfFan** [comment]
It appears the transform of `fopen` to `_fopen` is done by LLVM. Urrrrrgh.

**Comment by ProfFan** [comment]
And this strange list of function manglings: Some with `_`, some not
```
creating: renameat renameat
creating: renamex_np renamex_np
creating: renameatx_np renameatx_np
creating: clearerr clearerr
creating: fclose fclose
creating: feof feof
creating: ferror ferror
creating: fflush fflush
creating: fgetc fgetc
creating: fgetpos fgetpos
creating: fgets fgets
creating: fopen _fopen
creating: fprintf fprintf
creating: fputc fputc
creating: fputs _fputs
creating: fread fread
creating: freopen _freopen
creating: fscanf fscanf
creating: fseek fseek
creating: fsetpos fsetpos
creating: ftell ftell
creating: fwrite _fwrite
creating: getc getc
creating: getchar getchar
creating: gets gets
creating: perror perror
creating: printf printf
creating: putc putc
creating: putchar putchar
creating: puts puts
creating: remove remove
creating: rename rename
creating: rewind rewind
creating: scanf scanf
creating: setbuf setbuf
creating: setvbuf setvbuf
creating: sprintf sprintf
creating: sscanf sscanf
creating: tmpfile tmpfile
creating: tmpnam tmpnam
creating: ungetc ungetc
creating: vfprintf vfprintf
creating: vprintf vprintf
creating: vsprintf vsprintf
creating: ctermid ctermid
creating: fdopen _fdopen
creating: fileno fileno
creating: pclose pclose
creating: popen _popen
creating: __srget __srget
creating: __svfscanf __svfscanf
creating: __swbuf __swbuf
creating: __sputc __sputc
creating: flockfile flockfile
creating: ftrylockfile ftrylockfile
creating: funlockfile funlockfile
creating: getc_unlocked getc_unlocked
creating: getchar_unlocked getchar_unlocked
creating: putc_unlocked putc_unlocked
creating: putchar_unlocked putchar_unlocked
creating: getw getw
creating: putw putw
creating: tempnam _tempnam
creating: fseeko fseeko
creating: ftello ftello
creating: snprintf snprintf
creating: vfscanf vfscanf
creating: vscanf vscanf
creating: vsnprintf vsnprintf
creating: vsscanf vsscanf
creating: dprintf dprintf
creating: vdprintf vdprintf
creating: getdelim getdelim
creating: getline getline
creating: asprintf asprintf
creating: ctermid_r ctermid_r
creating: fgetln fgetln
creating: fmtcheck fmtcheck
creating: fpurge fpurge
creating: setbuffer setbuffer
creating: setlinebuf setlinebuf
creating: vasprintf vasprintf
creating: zopen zopen
creating: funopen funopen
```

**Comment by ProfFan** [comment]
6a1a119 should make it all green for merging (Hopefully)

File: src/tcompiler.cpp [hunk 5]
```diff
@@ -1304,7 +1363,13 @@ struct FunctionEmitter {
                         }
                         CU->fpm->run(*scc[i]);
                         VERBOSE_ONLY(T) {
+                            #if LLVM_VERSION < 38
```
Comment (zdevito) [inline src/tcompiler.cpp]: This pattern occurs in a few places and should probably be encapsulated in a function

Comment (ProfFan) [reply src/tcompiler.cpp]: Done

**Comment by ProfFan** [comment]
Glad that you liked it :)

Also, do you think it is appropriate to start using code linters like clang-tidy?

**Comment by ProfFan** [comment]
For the CUDA issue:
```
./terra tests/cudaatomic.t                                    ⬡ 8.11.4 [±master ●●]
-O3
-Wno-deprecated
-resource-dir
$CLANG_RESOURCE$
-I
.
-I
/usr/local/cuda/include
[cudalib] compile: ver 0
1596 terra -arch=compute_61

CUDA Module:
; ModuleID = 'terra'
source_filename = "terra"
target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v16:16:16-v32:32:32-v64:64:64-v128:128:128-n16:32:64"
target triple = "nvptx64-unknown-cuda"

; Function Attrs: alwaysinline
define void @bar(i32*) #0 {
entry:
  %t = alloca i32
  %result = alloca i32*
  store i32* %0, i32** %result
  %1 = call i32 @llvm.nvvm.read.ptx.sreg.tid.x()
  store i32 %1, i32* %t
  %2 = load i32*, i32** %result
  %3 = load i32, i32* %t
  call void asm sideeffect "red.global.max.u32 [$0], $1;", "l,r"(i32* %2, i32 %3)
  ret void
}

; Function Attrs: nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.tid.x() #1

attributes #0 = { alwaysinline }
attributes #1 = { nounwind readnone }

!nvvm.annotations = !{!0}

!0 = !{void (i32*)* @bar, !"kernel", i32 1}
Generated PTX:
//
// Generated by NVIDIA NVVM Compiler
//
// Compiler Build ID: CL-23757830
// Cuda compilation tools, release 9.2, V9.2.64
// Based on LLVM 3.4svn
//

.version 6.2
.target sm_61
.address_size 64




[cudalib] cuModuleGetFunction: bar
tests/cudaatomic.t:24: [cudalib] cuModuleGetFunction: cuda reported error 500: named symbol not found
stack traceback:
	[C]: in function 'error'
	[string "<string>"]:252: in function 'cudacompile'
	tests/cudaatomic.t:24: in main chunk
```
NVVM is generating empty output, thus no symbol :)

**Comment by ProfFan** [comment]
Tested on my Mac: with LLVM 5, CUDA 8.0, 9.0, 9.2 DNW; with LLVM 6, CUDA 8.0, 9.0, 9.2 DNW.

Symptom all the same

**Comment by elliottslaughter** [comment]
Thanks, I'm taking these changes into master since the basic functionality is working. I really appreciate your work on this, as it's been a sore spot on the project for a while.

If you'd like to work on CUDA as well, that would be great too, but no worries if you'd rather not.

I may see if I can get a clang-format config that follows the existing coding conventions, less sure about clang-tidy because I'm guessing the false positive rate might be high (but who knows, maybe it'll surprise me).

Thanks again for doing this!

**Comment by elliottslaughter** [comment]
Just FYI, I confirmed performance parity with LLVM 3.8 on the Regent program `examples/pennant_fast.rg`, which I think is a pretty reasonable test that performance hasn't regressed.

**Pull request merged**